vendor/bat-native-ledger/src/bat/ledger/internal/wallet_provider/transfer.cc

```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/wallet_provider/transfer.h"
+
+#include <utility>
+
+#include "base/guid.h"
+#include "base/strings/string_number_conversions.h"
+#include "bat/ledger/internal/ledger_impl.h"
+
+namespace ledger::wallet_provider {
+
+Transfer::Transfer(LedgerImpl* ledger) : ledger_(ledger) {
+  DCHECK(ledger_);
+}
+
+Transfer::~Transfer() = default;
+
+void Transfer::Run(const std::string& contribution_id,
+                   const std::string& destination,
+                   double amount,
+                   ledger::ResultCallback callback) const {
+  MaybeCreateTransaction(
+      contribution_id, destination,
+      // We can one-time/monthly tip in multiples of 0.25.
+      // A-C we can do in multiples of 1 (which is also a multiple of 0.25).
+      // The least amount one can tip is 0.25, where:
+      //   - 0.2375 (95%) is the actual contribution, and
+      //   - 0.0125 (== 1/80) (5%) is the fee
+      // Rounding to nearest ten-thousandth:
+      base::NumberToString(std::round(amount * 10000) / 10000),
+      base::BindOnce(&Transfer::CommitTransaction, base::Unretained(this),
+                     std::move(callback)));
+}
+
+void Transfer::MaybeCreateTransaction(
+    const std::string& contribution_id,
+    const std::string& destination,
+    const std::string& amount,
+    MaybeCreateTransactionCallback callback) const {
+  ledger_->database()->GetExternalTransaction(
+      contribution_id, destination,
+      base::BindOnce(&Transfer::OnGetExternalTransaction,
+                     base::Unretained(this), std::move(callback),
+                     contribution_id, destination, amount));
+}
+
+void Transfer::OnGetExternalTransaction(
+    MaybeCreateTransactionCallback callback,
+    std::string&& contribution_id,
+    std::string&& destination,
+    std::string&& amount,
+    absl::optional<mojom::ExternalTransactionPtr> transaction) const {
+  if (!transaction) {
+    return std::move(callback).Run(nullptr);
+  }
+
+  if (*transaction) {
```

Comment:
It feels odd to have to do this double-empty check. Should we just use `mojom::ExternalTransactionPtr` instead of `optional<mojom::ExternalTransactionPtr>` here? What are the benefits of wrapping with optional?

Reply:
`absl::nullopt` represents a failed database request, whereas `nullptr` means that there hasn't been an external transaction created for this contribution.

Reply:
I see. Are there any other options for that callback signature that would make it more self-documenting?

Reply:
So we need to be able to tell them apart, as we must not create a transaction on a failed lookup, but there aren't too many pleasing ways to do that. At least not that I can think of at the moment.
I'm usually against such signatures:
```
using GetExternalTransactionCallback =
    base::OnceCallback<void(bool success, mojom::ExternalTransactionPtr)>;
```
as an `absl::optional<T>` expresses the intent much clearer in my opinion. It just happens to look a bit weird, since here `T` is a pointer type, which already has optional semantics.

Reply:
I see your point about `optional` expressing the semantics better than boolean. Would `expected` be another option? If we used `expected` then we could have two separate error states: one for database access error and one for "record not found", and the transaction object would always be non-empty if there's no error. Is that a possibility?

Reply:
Intuitively, it seems like both paths that don't result in a transaction record are errors, since we've asked the database for a record and it couldn't give us one. If we model those two cases as an error enum:

```cpp

enum class GetExternalTransactionError {
  kDatabaseError = 1,
  kTransactionNotFound
};

```

then would one of these work?

```cpp

// 1.

using GetExternalTransactionCallback =
    base::OnceCallback<void(base::expected<mojom::ExternalTransactionPtr,
                                           GetExternalTransactionError>)>;

// 2.

using GetExternalTransactionCallback =
    base::OnceCallback<void(GetExternalTransactionError,
                            mojom::ExternalTransactionPtr)>;

```

Reply:
`1.` would, but the `kTransactionNotFound` case is not an error, therefore using `base::expected<>` seems a bit artificial to me.

Reply:
One could view it as a 404-like error, no?

Reply:
Hmm, I don't think so? 🙂 I mean, `HTTP 404` usually refers to an error which is the result of trying to follow a dead link. There the user expects to find something at the address. This is not the case here, as not having a transaction in the database only means that the contribution engine is running the contribution for the first time, so it has to create one.

Reply:
It's incorrect to assume that there's already a transaction connected to the contribution.

Reply:
Not sure if that's better, or worse, but let's make everybody happy! 😄 https://github.com/brave/brave-core/pull/16444/commits/7e45e95f0f18989ed6b711fbe4d9fce7bb6f76ee

Reply:
Is 204 (No Content) status code useful here, @szilardszaloki ?